### PR TITLE
test(undici): add interceptor tests for undici

### PR DESCRIPTION
**Pull request by chentsulin**
See #159.

### CODE REVIEW

File: test/intercept/undici/undici.fetch.test.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
+/**
```
Comment (kettanaito) [inline test/intercept/undici/undici.fetch.test.ts:1]: We can put these test suites into `test/integration` folder. It's there to test this library's integration with other libraries (like `got`), as those may have certain corner-cases to account for.

Comment (chentsulin) [reply test/intercept/undici/undici.fetch.test.ts:1]: As I see it, `undici` is a replacement for the old `http` and `https` module, and many libraries [include got](https://github.com/sindresorhus/got/issues/1419) may use it instead of `http` in the future. So, I suppose this test should be placed nearby the tests of the `http` interceptor.

Comment (kettanaito) [reply test/intercept/undici/undici.fetch.test.ts:1]: @chentsulin, well, `undici` is still a _third-party library_. Until it's a part of the Node.js standard library, we should treat it as third-party integration.

**Comment by kettanaito** [comment]
I've rebased your feature branch, as well as adjusted it to the latest test structure refactoring (#182). Your newly added undici tests are now under `test/third-party/undici`.

File: test/intercept/undici/undici.fetch.test.ts (hunk 1 above)
Comment (chentsulin) [reply test/intercept/undici/undici.fetch.test.ts]: Fair enough. We can rethink about that by then.

**Comment by kettanaito** [comment]
Thank you for adding these tests!

The support for unidici will roughly follow these steps:

1. Research how `undici` performs requests internally. See what modules it utilizes, what modules can be extended to support interception.
2. Discuss the findings, assess the support strategy.
2. If agreed, create a new `interceptors/undici` interceptor with the barebone implementation to cover the tests you've added.
3. Ensure sufficient test coverage, merge and release the change. 

If you have time, feel free to tackle the first step. Gathering knowledge often takes the most time, and I could definitely use your expertise in that. Thank you!

**Comment by chentsulin** [comment]
In my early research, undici has its own mocking mechanism:

```js
import { MockAgent, setGlobalDispatcher, } from 'undici'

const mockAgent = new MockAgent();

setGlobalDispatcher(mockAgent);

// Provide the base url to the request
const mockPool = mockAgent.get('http://localhost:3000');

// intercept the request
mockPool.intercept({
  path: '/bank-transfer',
  method: 'POST',
  headers: {
    'X-TOKEN-SECRET': 'SuperSecretToken',
  },
  body: JSON.stringify({
    recepient: '1234567890',
    ammount: '100'
  })
}).reply(200, {
  message: 'transaction processed'
})
```

We should take advantage of this [setGlobalDispatcher](https://github.com/nodejs/undici#undicisetglobaldispatcherdispatcher) stuff to implement the interceptor.

**Comment by chentsulin** [comment]
@kettanaito,
I don't have enough knowledge to experiment and finish all implementation details, but I do a naive and maybe buggy implementation for you to reference:
https://github.com/mswjs/interceptors/commit/a7e21d1a929e0914cdd472e73ac68bdce676c75a

That's what i learned:

- `setGlobalDispatcher` and `setGlobalDispatcher` are prefect for mocking and restoring.
- undici classes are written in ES2015, so set `compilerOptions.target` to `ES5` won't work.
- The `dispatch` method is synchronous, so have to find a way to inject asynchronous body parsing and response resolving.
- Relying on `dispatch` and `handler` seem to be right, because undici use them: https://github.com/nodejs/undici/blob/f951002dfb86f502f1129588c021a26d7f2a0afb/lib/mock/mock-utils.js#L168-L174

**Comment by jtlopez18** [comment]
@chentsulin Thanks for sharing your implementation, it was super helpful.  I have been playing with it and seeing an issue where undici expects @types/node to be at least 14.18 (this has the Blob type from buffer) in order to import the Dispatcher.DispatchHandlers type for the resolver.

However, updating this seems to cause issues with other intercepters, such as XML and http.get.  Did you run into anything like that?  How did you resolve?


> @kettanaito, I don't have enough knowledge to experiment and finish all implementation details, but I did a naive and maybe buggy implementation for you to refer: [a7e21d1](https://github.com/mswjs/interceptors/commit/a7e21d1a929e0914cdd472e73ac68bdce676c75a)

**Comment by kettanaito** [comment]
Thanks for preparing a reference implementation, @chentsulin! 

I'm not sure to which extend we can reuse it, as with interceptors we're aiming at low-level request interception. Relying on `MockAgent` is a high-level thing, meaning we're going deeper into implementation specifics rather than relying on Node.js primitives. 

I'd love to start this feature from research. We should first understand what kind of requests does undici make (it _has_ to make a `Socket` connection at least, there are no other ways to request things in Node.js) and, ideally, model our interceptor around that. 

I don't think I'll have time to work on this in the foreseeable future, so anybody is welcome to grab the research and implementation if they find this interesting.
